vkcs/backup/resource_vkcs_backup_plan.go

```diff
@@ -0,0 +1,597 @@
+					},
+				},
+				Validators: []validator.Object{
+					objectvalidator.ConflictsWith(path.Expressions{
+						path.MatchRoot("full_retention"),
+					}...),
+				},
+				Description: "Parameters for gfs retention policy. Incompatible with full_retention",
+			},
+
+			"incremental_backup": schema.BoolAttribute{
+				Required:    true,
+				Description: "Whether incremental backups should be stored",
+			},
+
+			"provider_id": schema.StringAttribute{
+				Required:    true,
+				Description: "ID of backup provider",
+			},
+
+			"instances_id": schema.ListAttribute{
+				ElementType: types.StringType,
+				Required:    true,
+				Description: "List of ids of instances to make backup for",
+			},
+
+			"region": schema.StringAttribute{
+				Computed:    true,
+				Optional:    true,
+				Description: "The `region` to fetch availability zones from, defaults to the provider's `region`.",
+			},
+		},
+		Description: "Manages a backup plan resource.",
+	}
+}
+
+func (r *PlanResource) Configure(ctx context.Context, req resource.ConfigureRequest, resp *resource.ConfigureResponse) {
+	if req.ProviderData == nil {
+		return
+	}
+	r.config = req.ProviderData.(clients.Config)
+}
+
+func (r *PlanResource) Create(ctx context.Context, req resource.CreateRequest, resp *resource.CreateResponse) {
+	var plan PlanResourceModel
+	diags := req.Plan.Get(ctx, &plan)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	region := plan.Region.ValueString()
+	if region == "" {
+		region = r.config.GetRegion()
+	}
+
+	backupClient, err := r.config.BackupV1Client(region, r.config.GetTenantID())
+	if err != nil {
+		resp.Diagnostics.AddError("Error creating VKCS backup client", err.Error())
+		return
+	}
+
+	resourcesInfo, err := getResourcesInfo(r.config, region, plan.InstancesID, plan.ProviderID.ValueString())
+	if err != nil {
+		resp.Diagnostics.AddError("Error creating vkcs_backup_plan", err.Error())
+		return
+	}
+
+	planCreateOpts := plans.CreateOpts{
+		Name:       plan.Name.ValueString(),
+		Resources:  resourcesInfo,
+		ProviderID: plan.ProviderID.ValueString(),
+	}
+
+	incrementalBackups := plan.IncrementalBackup.ValueBool()
+	if incrementalBackups {
+		fullDay, err := getFullDayFromPlan(plan)
+		if err != nil {
+			resp.Diagnostics.AddError("Error creating vkcs_backup_plan", err.Error())
+			return
+		}
+		planCreateOpts.FullDay = fullDay
+	}
+
+	if plan.FullRetention != nil {
+		planCreateOpts.RetentionType = RetentionFull
+	} else if plan.GFSRetention != nil {
+		planCreateOpts.RetentionType = RetentionGFS
+		planCreateOpts.GFS = getGFSFromPlan(plan)
+	} else {
+		resp.Diagnostics.AddError("Invalid resource schema", "One of full_retention, gfs_retention must be specified")
+		return
+	}
+
+	planResp, err := plans.Create(backupClient, &plans.PlanCreate{Plan: &planCreateOpts}).Extract()
+	if err != nil {
+		resp.Diagnostics.AddError("Error creating vkcs_backup_plan", err.Error())
+	}
+	planID := planResp.ID
```

Comment:
It would be better to set id to state here because any leading error will cause to having an "unmanaged" remote object, and setting id will allow Terraform to mark it as tainted.

Reply:
Moved setting id here

Reply:
I think we don't need such precaution here because resource is created in backend immediately, and we do not wait for it to reach certain status. Though, we may set id here for the purposes of similarity with other resources

Reply:
You are right that there is no waiting for resource creation here, but leading code can raise an error. For example, a plan object has been created but user provides a wrong configuration for trigger schedule, and we return error. Thus, there is no state for vkcs_backup_plan because we haven't set anything in there, and object cannot be destroyed because we don't know that it exists, and user can't even recreate it as recreation will lead for duplicate backup plan error.

Reply:
Agree, rewrote to set right to state

Reply:
Also, may be we should consider at-plan validation for schedule time configuration here, than there should be no errors in getting triggerschedule. Also, error at line 294 may be misleading, as it seems to be unrelated to retentions. There is also validation helper for checking if at least one of attribute is set: https://pkg.go.dev/github.com/hashicorp/terraform-plugin-framework-validators@v0.10.0/resourcevalidator#AtLeastOneOf. Also, methods that transform state into API objects are commonly called expandSomething (like flattenSomething) in providers, but it's up to you, this name is also ok.